File: src/appearance.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial } from "./game.js";
import { featureStep } from "./features.js";
import { ArenaAudio } from "./audio.js";
import { notificationEvent } from "./events.js";
test("each item announces appearance once, and replayed snapshots do not repeat it", () => {
 const s=initial();for(const p of s.players)p.stun=100;
 for(const item of s.pickups)item.wait=.08;
 featureStep(s,.04);assert.equal(s.events?.length??0,0);
 featureStep(s,.04);
 assert.equal(s.events.length,7);
 s.events.forEach((e,i)=>{assert.equal(e.kind,21);assert.equal(e.x,s.pickups[i].x);assert.equal(e.z,s.pickups[i].z)});
 assert.equal(notificationEvent(s,0),null);
 const audio=new ArenaAudio(),heard=[];audio.play=k=>heard.push(k);
 audio.observe(s,true);audio.observe(structuredClone(s),true);
 featureStep(s,.04);audio.observe(s,true);
 assert.equal(heard.filter(k=>k===21).length,7);
});
test("cash limit suppresses appearance audio and held gear remains silent", () => {
 const s=initial();for(const p of s.players)p.stun=100;
 for(const item of s.pickups)item.wait=100;
 s.credits=[2000,2000];s.pickups[2].wait=.04;
 s.pickups[6].kind=0;s.pickups[6].wait=.04;
 featureStep(s,.04);assert.equal(s.events?.length??0,0);
});

File: server/appearance_test.go
package main

import "testing"

func TestItemAppearanceEvents(t *testing.T) {
	s := initial()
	for i := range s.Players {
		s.Players[i].Stun = 100
	}
	for i := range s.Pickups {
		s.Pickups[i].Wait = .08
	}
	s.featureStep(.04)
	if s.Event.ID != 0 {
		t.Fatal("early announcement")
	}
	s.featureStep(.04)
	if s.Event.ID != 7 || s.Event.Kind != 21 {
		t.Fatal("appearance events", s.Event)
	}
	s.featureStep(.04)
	if s.Event.ID != 7 {
		t.Fatal("repeated appearance")
	}
}
func TestSuppressedAppearanceIsSilent(t *testing.T) {
	s := initial()
	for i := range s.Players {
		s.Players[i].Stun = 100
	}
	for i := range s.Pickups {
		s.Pickups[i].Wait = 100
	}
	s.Credits = [2]int{2000, 2000}
	s.Pickups[2].Wait = .04
	s.Pickups[6].Kind = 0
	s.Pickups[6].Wait = .04
	s.featureStep(.04)
	if s.Event.ID != 0 {
		t.Fatal("hidden item made sound")
	}
}
